openshift/osd-network-verifier
# Allow user to choose between multiple endpoint lists at runtime

## What does this PR do?
This PR addresses [OSD-15143](https://issues.redhat.com/browse/OSD-15143), which calls for conditional logic to support hypershift-specific endpoints. We've decided to generalize this to "support any arbitrary list of endpoints baked into the golden AMI." To accomplish this, this PR adds the string field `PlatformType` to the `ValidateEgressInput` struct (and a corresponding `--platform` CLI argument), allowing the user to specify which infrastructure platform the subnet-under-test is intended for (either `aws`, `gcp`, or `hostedcluster` (hypershift 'child' cluster)). This in turn specifies which egress endpoint config file (that's baked into the golden AMI) should be used (e.g., `hostedcluster.yaml`). This field is backwards-compatible (i.e., optional) and will default to `aws` if no value is provided.

This PR focuses on the AWS/Hypershift implementations. GCP implementation will be tackled in a future card/PR.

**Compatibility Warning**: this PR should not be merged until a new golden AMI is published containing three new files in the `/build/config` directory: `aws.yaml`, `gcp.yaml`, and `hostedcluster.yaml`. I've provided a patch in the testing instructions below to get around this incompatibility in the meantime.

## Checklist

## How to test this PR
1. Checkout this PR and `cd` into the root of the repo.
2. ~Run the following command to apply [this patch](https://gist.github.com/abyrne55/8b865fe0dcd96571b7bc263339718d1a), which will configure your build to use a [testing docker image](https://quay.io/repository/abyrne_openshift/osd-network-verifier/manifest/sha256:692b3262d42f31b67612e66e68ea14c68090a4055adb3287a67bd419d2595a07) loaded with dummy configurations (see these in [my golden-ami fork](https://gitlab.cee.redhat.com/abyrne/osd-network-verifier-golden-ami/-/tree/OSD-15143/build/config) using RH-VPN): `curl https://gist.githubusercontent.com/abyrne55/8b865fe0dcd96571b7bc263339718d1a/raw/de74e1a64cfdc208d1a2f6323407fb12801d22b7/OSD-15143.use-dummy-ami.patch | git apply --index`~ _This step is no longer necessary now that real configs are present in the golden AMI_
3. Run `make build`
4. Run the verifier as you normally would, adding the `--platform` argument with values `aws` or `hostedcluster` (or leave the arg off and ensure it still defaults to AWS), e.g.
```shell
./osd-network-verifier egress --platform=aws --subnet-id=$SUBNET_ID --security-group-id=$SEC_GRP_ID --region=$AWS_REGION --profile=$AWS_PROFILE --debug 
```
5. ~Ensure that at least one failure is reported for the fake domain `thisis.<platform_arg_value>.yaml:80`~ This step no longer applies when using the real configs that recently landed in the golden AMI repo. Instead, set up your test security group to block access to endpoints not needed by hosted clusters (e.g., `nosnch.in`), and ensure it shows up as an egress failure for `--platform=aws` but not for `--platform=hostedcluster`

## Logs
Example output for `--platform=hostedcluster`
```
$ ./osd-network-verifier egress --subnet-id=subnet-00b7c4c20ee8c422a --security-group-id=sg-01d674f6f338b5c67 --region=us-east-1 --profile=default --platform=hostedcluster
Using region: us-east-1
Created instance with ID: i-035250618e62296ce
Summary:
printing out failures:
 -  egressURL error: Unable to reach tagging.us-east-1.amazonaws.com:80
 -  egressURL error: Unable to reach thisis.hostedcluster.yaml:80
 -  egressURL error: Unable to reach events.amazonaws.com:443
 -  egressURL error: Unable to reach inputs1.osdsecuritylogs.splunkcloud.com:9997
 -  egressURL error: Unable to reach servicequotas.us-east-1.amazonaws.com:80
Failure!
```
